### ISSUE TITLE
DNN: Access violation in deconv2d layer

### ISSUE THREAD

**Issue by philipp-fischer** [opencv/opencv#10702]
[minimal_net.py.txt](https://github.com/opencv/opencv/files/1668899/minimal_net.py.txt)

##### System information (version)
- OpenCV => 3.4 (Master Branch) 
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015

##### Detailed description
Exception in deconvolution layer.
Easily reproducible with minimal example (see attachments) 
Access violation happens im "convolution_layer.cpp" line 1089.
(in the two for loops "val +=...") 

##### Steps to reproduce
Set up a minimal one-layer tensorflow network as in the attached file and export it.
Remove unneeded ops using the tf optimizer. 
Load the protobuf into opencv with readNetFromTensorflow and do a forward pass. 
It will crash. 
![20180126_200027](https://user-images.githubusercontent.com/12006597/35456341-61e96fa6-02d5-11e8-9a75-ad3f728a97f1.png)
[minimal_net.py.txt](https://github.com/opencv/opencv/files/1668900/minimal_net.py.txt)

[deconv.zip](https://github.com/opencv/opencv/files/1668978/deconv.zip)

**Comment by akhandait** [comment]
@dkurt I would like to work on this issue, can you give me some initial directions to get started?

**Comment by philipp-fischer** [comment]
Great. You can build open this example:

https://github.com/opencv/opencv/blob/master/samples/dnn/tf_inception.cpp

But you can really strip it down to
readNetFromTensorflow
SetInput and net.forward. 

I attached a zip file that contains the optimized_model.pb you can load.

**Comment by dkurt** [comment]
@akhandait, as @philipp-fischer said firstly we need to reproduce a error. Download attached graph, try to run it using OpenCV.
Then we need to find a problem. There is a [generate_tf_models.py](https://github.com/opencv/opencv_extra/blob/master/testdata/dnn/tensorflow/generate_tf_models.py) script to generate test data using TensorFlow. We already have one test for `conv2d_transpose` so we need to figure out what configuration leads the crush. I think you can start change strides and padding arguments then change the rest of parameters until you'll catch a bug.

**Comment by akhandait** [comment]
@dkurt @philipp-fischer Thanks, I will get on it now.

**Comment by philipp-fischer** [comment]
I had a look at the `generate_tf_models.py` script.
The conv2d_transpose layer there only uses strides of 1 and VALID padding.
Maybe you can already reproduce the error by changing this to stride 2 and SAME padding as a second test case.

**Comment by philipp-fischer** [comment]
I am also wondering why opencv doesn't seem to load the output_size from the TF graph. I think it is crucial for the layer to work.

**Comment by akhandait** [comment]
I stripped the tf_inception.cpp as you said.
Here it is.
`
#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
using namespace cv;
using namespace cv::dnn;

#include <fstream>
#include <iostream>
#include <cstdlib>
using namespace std;

const String keys =
        "{help h    || Sample app for loading Inception TensorFlow model. "
                       "The model and class names list can be downloaded here: "
                       "https://storage.googleapis.com/download.tensorflow.org/models/inception5h.zip }"
        "{model m   |tensorflow_inception_graph.pb| path to TensorFlow .pb model file }"
        "{image i   || path to image file }"
        "{i_blob    | input | input blob name) }"
        "{o_blob    | softmax2 | output blob name) }"
        "{result r  || path to save output blob (optional, binary format, NCHW order) }"
        ;

void getMaxClass(const Mat &probBlob, int *classId, double *classProb);
std::vector<String> readClassNames(const char *filename);

int main(int argc, char **argv)
{
    cv::CommandLineParser parser(argc, argv, keys);

    if (parser.has("help"))
    {
        parser.printMessage();
        return 0;
    }

    String modelFile = parser.get<String>("model");
    String imageFile = parser.get<String>("image");
    String inBlobName = parser.get<String>("i_blob");
    String outBlobName = parser.get<String>("o_blob");

    if (!parser.check())
    {
        parser.printErrors();
        return 0;
    }

    String resultFile = parser.get<String>("result");

    //! [Initialize network]
    dnn::Net net = readNetFromTensorflow(modelFile);
    //! [Initialize network]

    if (net.empty())
    {
        std::cerr << "Can't load network by using the mode file: " << std::endl;
        std::cerr << modelFile << std::endl;
        exit(-1);
    }

    //! [Prepare blob]
    Mat img = imread(imageFile);
    if (img.empty())
    {
        std::cerr << "Can't read image from the file: " << imageFile << std::endl;
        exit(-1);
    }

    Mat inputBlob = blobFromImage(img, 1.0f, Size(32, 32), Scalar(), true, false); 
    //Convert Mat to batch of images
    //! [Prepare blob]
    //! [Set input blob]
    net.setInput(inputBlob, inBlobName);        //set the network input
    //! [Set input blob]
    cv::TickMeter tm;
    tm.start();

    //! [Make forward pass]
    Mat result = net.forward(outBlobName);                          //compute output
    //! [Make forward pass]
    tm.stop();

    if (!resultFile.empty()) {
        CV_Assert(result.isContinuous());

        ofstream fout(resultFile.c_str(), ios::out | ios::binary);
        fout.write((char*)result.data, result.total() * sizeof(float));
        fout.close();
    }

    std::cout << "Output blob shape " << result.size[0] << " x " << result.size[1] << " x " << result.size[2] << " x " << result.size[3] << std::endl;
    std::cout << "Inference time, ms: " << tm.getTimeMilli()  << std::endl;
    
    return 0;
} `

But, I ran into a failed assertion and haven't been able to reproduce the bug. Can you help me out a bit here, I think it's about the wrong input size of the blob.
 
OpenCV Error: Assertion failed (blobs[0].size[0] == inpCn) in getMemoryShapes, file /home/atharva/Documents/opencv/modules/dnn/src/layers/convolution_layer.cpp, line 1070
terminate called after throwing an instance of 'cv::Exception'
  what():  /home/atharva/Documents/opencv/modules/dnn/src/layers/convolution_layer.cpp:1070: error: (-215) blobs[0].size[0] == inpCn in function getMemoryShapes

Aborted (core dumped)

**Comment by akhandait** [comment]
I stripped tf_inception.cpp down to what we need.
Here it is:
``` C++
#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
using namespace cv;
using namespace cv::dnn;

#include <fstream>
#include <iostream>
#include <cstdlib>
using namespace std;

const String keys =
        "{help h    || Sample app for loading Inception TensorFlow model. "
                       "The model and class names list can be downloaded here: "
                       "https://storage.googleapis.com/download.tensorflow.org/models/inception5h.zip }"
        "{model m   |tensorflow_inception_graph.pb| path to TensorFlow .pb model file }"
        "{image i   || path to image file }"
        "{i_blob    | input | input blob name) }"
        "{o_blob    | softmax2 | output blob name) }"
        "{result r  || path to save output blob (optional, binary format, NCHW order) }"
        ;

void getMaxClass(const Mat &probBlob, int *classId, double *classProb);
std::vector<String> readClassNames(const char *filename);

int main(int argc, char **argv)
{
    cv::CommandLineParser parser(argc, argv, keys);

    if (parser.has("help"))
    {
        parser.printMessage();
        return 0;
    }

    String modelFile = parser.get<String>("model");
    String imageFile = parser.get<String>("image");
    String inBlobName = parser.get<String>("i_blob");
    String outBlobName = parser.get<String>("o_blob");

    if (!parser.check())
    {
        parser.printErrors();
        return 0;
    }

    String resultFile = parser.get<String>("result");

    //! [Initialize network]
    dnn::Net net = readNetFromTensorflow(modelFile);
    //! [Initialize network]

    if (net.empty())
    {
        std::cerr << "Can't load network by using the mode file: " << std::endl;
        std::cerr << modelFile << std::endl;
        exit(-1);
    }

    //! [Prepare blob]
    Mat img = imread(imageFile);
    if (img.empty())
    {
        std::cerr << "Can't read image from the file: " << imageFile << std::endl;
        exit(-1);
    }

    Mat inputBlob = blobFromImage(img, 1.0f, Size(32, 32), Scalar(), true, false); 
    //Convert Mat to batch of images
    //! [Prepare blob]
    //! [Set input blob]
    net.setInput(inputBlob, inBlobName);        //set the network input
    //! [Set input blob]
    cv::TickMeter tm;
    tm.start();

    //! [Make forward pass]
    Mat result = net.forward(outBlobName);                          //compute output
    //! [Make forward pass]
    tm.stop();

    if (!resultFile.empty()) {
        CV_Assert(result.isContinuous());

        ofstream fout(resultFile.c_str(), ios::out | ios::binary);
        fout.write((char*)result.data, result.total() * sizeof(float));
        fout.close();
    }

    std::cout << "Output blob shape " << result.size[0] << " x " << result.size[1] << " x " << result.size[2] << " x " << result.size[3] << std::endl;
    std::cout << "Inference time, ms: " << tm.getTimeMilli()  << std::endl;
    
    return 0;
} 
```

But, I ran into a failed assertion and haven't been able to reproduce the bug. Can you help me out here a bit? I think it's about the wrong input blob size.
```
OpenCV Error: Assertion failed (blobs[0].size[0] == inpCn) in getMemoryShapes, file /home/atharva/Documents/opencv/modules/dnn/src/layers/convolution_layer.cpp, line 1070
terminate called after throwing an instance of 'cv::Exception'
  what():  /home/atharva/Documents/opencv/modules/dnn/src/layers/convolution_layer.cpp:1070: error: (-215) blobs[0].size[0] == inpCn in function getMemoryShapes

Aborted (core dumped)
```

**Comment by dkurt** [comment]
@philipp-fischer, you are right. It's the next problem but it's not critical for all cases. In example, 
```python
inp = tf.placeholder(tf.float32, [1, 31, 31, 1], 'input')
weights = tf.Variable(tf.random_normal([3, 3, 8, 1], dtype=tf.float32))
deconv = tf.nn.conv2d_transpose(value=inp, filter=weights,
                                output_shape=[1, 63, 63, 8], strides=[1, 2, 2, 1],
                                padding='VALID')
```
works fine because it's exactly inverse to convolution layer, `31 = floor((63 - 3) / 2) + 1`. But if `output_shape=[1, 64, 64, 8]` we have to add some `adjust` padding (here it's 1x1). It's already implemented for [Torch's deconvolution](https://github.com/torch/nn/blob/master/doc/convolution.md#spatialfullconvolution) and should be computed for TensorFlow from `output_shape` as you noticed.

**Comment by dkurt** [comment]
@akhandait, we have more convenient way for tests. Take a look on [test_tf_importer.cpp](https://github.com/opencv/opencv/blob/master/modules/dnn/test/test_tf_importer.cpp). There is a test
```cpp
TEST(Test_TensorFlow, deconvolution)
{
    runTensorFlowNet("deconvolution");
}
```
that could be extended with one more test case. You need to generate input and output for network so TensorFlow should be installed first. Then save model, input and target output at `opencv_extra/testdata/dnn` and run a test by `opencv_test_dnn --gtest_filter=Test_TensorFlow.deconvolution`.

**Comment by philipp-fischer** [comment]
Had some quick tests today. So it seems to crash only if you use strides larger than 1.
VALID vs. SAME padding didn't cause a crash for me.

**Comment by akhandait** [comment]
@philipp-fischer In the generate_tf_models.py file, we have to provide an argument for the transform_graph tool, which file does it expect? I searched for it in my installed tensorflow but couldn't find it anywhere. How did you run it?

**Comment by dkurt** [comment]
@akhandait, [Using the Graph Transform Tool](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#using-the-graph-transform-tool)

**Comment by dkurt** [comment]
> Had some quick tests today. So it seems to crash only if you use strides larger than 1.
VALID vs. SAME padding didn't cause a crash for me.

@philipp-fischer, I think we could face a problem varying both strides or padding mode. In example, strides are 1 but `SAME` padding is crashed:
```python
inp = tf.placeholder(tf.float32, [1, 64, 64, 1], 'input')
deconv_weights = tf.Variable(tf.random_normal([3, 3, 8, 1], dtype=tf.float32), name='deconv_weights')
deconv = tf.nn.conv2d_transpose(value=inp, filter=deconv_weights,
                                output_shape=[1, 64, 64, 8], strides=[1, 1, 1, 1],
                                padding='SAME')
```

but the following example works:
```python
inp = tf.placeholder(tf.float32, [1, 64, 64, 1], 'input')
deconv_weights = tf.Variable(tf.random_normal([3, 3, 8, 1], dtype=tf.float32), name='deconv_weights')
deconv = tf.nn.conv2d_transpose(value=inp, filter=deconv_weights,
                                output_shape=[1, 66, 66, 8], strides=[1, 1, 1, 1],
                                padding='VALID')
```

**Comment by philipp-fischer** [comment]
I think I found it:
The problem is in `class BaseConvolutionLayerImpl` in `void finalize(...)`.
This function is used for both conv and deconv layers when calling `getConvPoolPaddings`.

But for deconv, the order of input and output should be the other way around when computing the padding.

My current workaround is to override the `finalize()` function in `DeConvolutionLayerImpl` and change the `getConvPoolPaddings` call by swapping the first two arguments.

**Comment by dkurt** [comment]
@philipp-fischer, May I ask you to test your model with changes from https://github.com/opencv/opencv/pull/10850 ?

**Comment by philipp-fischer** [comment]
Thanks! We had a quick test and it seems to work now.
Great that you even fixed the shape problems with `adj_w` and `adj_h`
